evm_loader/program/src/executor/state.rs

```diff
@@ -285,17 +285,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
     }
 
     async fn code_hash(&self, from_address: Address, chain_id: u64) -> Result<[u8; 32]> {
-        use solana_program::keccak::hash;
-
-        for action in &self.actions {
-            if let Action::EvmSetCode { address, code, .. } = action {
-                if &from_address == address {
-                    return Ok(hash(code).to_bytes());
-                }
-            }
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // https://eips.ethereum.org/EIPS/eip-161
+
+        #[maybe_async]
+        async fn data_account_exists<B: AccountStorage>(
+            state: &ExecutorState<'_, B>,
+            address: Address,
+            chain_id: u64,
+        ) -> Result<bool> {
+            Ok(state.nonce(address, chain_id).await? > 0
+                || state.balance(address, chain_id).await? > 0)
         }
 
-        Ok(self.backend.code_hash(from_address, chain_id).await)
+        // FIXME: Can we modify self.code to return Option<Buffer> or Option<&[u8]>?
+
+        // FIXME: Because buffer is returned by value, we need to store buffer in order to store a
+        // reference to the bytes. I could use Option<Buffer> instead, but that is storing a more
+        // powerful type than I need. I just need a byte slice.
```

Comment:
I think `Database::code` method should still return just `Buffer`, because it is hard to keep the invariant that `Option<Buffer>` contains a non-empty `Buffer`. It can be checked if needed.

Reply:
I need to know whether a program account exists for that address. I does not matter if the code for the account for it is empty, whether that is technically allowed or not. I think this is best communicated through an `Option<Buffer>` or even a `Result<Buffer, AccountDoesNotExistError>`. If a program accounts always has a non empty number of code bytes (I would say this is probably the case, but not necessarily), then we **can** also represent the non-existance through `Vec::empty`. However, is that a good choice from a maintainability point of view? Would that encourage consumers of the method to check for this when it matters?

Reply:
Not sure if you posted this in the right comment. The question we need to resolve here is: "what should the function signature of `code` look like?".

Reply:
Keeping it as `Buffer` because that is what most methods do right now: to return a sentinel value (`0`, empty vec, ...) instead of `None`.